managed-pointer/cpp-20/D0567R1.md

```diff
@@ -229,94 +391,105 @@ auto gpuExecutor = gpuContext.executor();
+
+## 4. Use Cases
 
-*Figure 4: Example of implicit interface*
+// TODO
 
-## Future Work
+## 5. Future Work
 
 There are many other considerations to make when looking at a model for data movement for heterogeneous and distributed systems, however, this paper aims to establish a foundation which can be extended to include other paradigms in the future.
 
-### More Complex Data Movement Policies
+### 5.1. Shared Accessibility
 
-We may wish to introduce more control over the way in which data is moved between devices at a high level and the consistency between execution contexts. This could be done by having static type traits associated with executor classes which describe the data movement and consistency properties of the execution context associated with the executor.
+Systems which support a unified address space (i.e. sharing a single memory region across different *execution contexts*) could allow for a `managed_ptr` instance to share accessibility between multiple *execution contexts*. This would allow a `managed_ptr` to be concurrently accessible by multiple *execution contexts* via the use of atomics providing those *execution contexts* share a memory region.
 
-### More Complex Execution Ordering Policies
+The current proposed solution is that the `managed_ptr` have an additional template parameter specifying the `memory_mode`, allowing it to be specialised for supporting different memory architectures. The `memory_mode` of a `managed_ptr` can be either `memory_mode::discrete`, `memory_mode::unified` or `memory_model::shared`:
 
-We may wish to introduce more control over the order in which operations (both data movement and compute) are executed. This could be done by having static type traits associated with executor classes which describe the ordering guarantees between operations, so rather than always having a strict sequential ordering, an implementation may want to relax the requirements based on access dependencies to allow for optimisations.
+* The memory mode `memory_mode::discrete` can only have a single accesible execution context so cannot be accessed concurrently and has discrete emory regions so cannot support passing pointer based structures.
+* The memory mode `memory_mode::unified` can only have a single accesible execution context so cannot be accessed concurrently but has a unified memory region so can support passing pointer based structures.
+* The memory mode `memory_mode::shared` can have multiple accesible execution contexts so can be accessed concurrently via the use of atomics and has a unified memory region so can support passing pointer based structures.
 
-### Synchronisation Operations From Multiple Host CPU Devices
+This proposal is not yet in the main paper and so is considered future work as the idea has not yet been fully fleshed out and there are other factors that have to be considered such as the execution context - memory region topology and atomic operations.
```

Comment:
I meant here that this idea has not been integrated into the proposal yet. I've re-worded this.